zeus/device/gpu.py

```diff
@@ -0,0 +1,908 @@
+        if (visible_device := os.environ.get("ROCR_VISIBLE_DEVICES")) is not None:
+            self.visible_indices = [int(idx) for idx in visible_device.split(",")]
+        else:
+            self.visible_indices = list(
+                range(len(amdsmi.amdsmi_get_processor_handles()))
+            )
+
+        # initialize all GPUs
+        self.gpus = {}
+        if gpus_to_track is not None:
+            for gpu_num in gpus_to_track:
+                self.gpus[gpu_num] = AMDGPU(self.visible_indices[gpu_num])
+        else:
+            for index, gpu_num in enumerate(self.visible_indices):
+                self.gpus[index] = AMDGPU(gpu_num)
+
+    def __init__(
+        self, gpus_to_track: list[int] = None, ensure_homogeneous: bool = True
+    ) -> None:
+        """Initializes the AMDGPUs instance, setting up tracking for specified AMD GPUs.
+
+        This involves initializing the amdsmi library to communicate with the amdsmi driver and configuring which GPUs
+        are visible based on the `ROCR_VISIBLE_DEVICES` environment variable and the optional `gpus_to_track`
+        parameter. It ensures that only the specified GPUs (if any) are tracked and managed by this instance.
+
+        Parameters:
+            gpus_to_track (list[int], optional): Specifies the indices of the GPUs to be tracked. These indices
+                should align with the visible GPU indices as determined by the system and the `ROCR_VISIBLE_DEVICES`
+                environment variable. If None, all visible GPUs are tracked.
+            ensure_homogeneous (bool, optional): If True, attempts to ensure that all tracked GPUs are of the same model
+                and specifications.
+
+        Raises:
+            ZeusBaseGPUError: If an amdsmi error occurs, a specific GPU error is raised,
+                encapsulating the original amdsmi error message for clearer debugging and error handling.
+        """
+        try:
+            amdsmi.amdsmi_init()
+            self._init_gpus(gpus_to_track)
+            if ensure_homogeneous:
+                self._ensure_homogeneous()
+        except amdsmi.AmdSmiException as e:
+            exception_class = AMDGPU._exception_map.get(e.value, ZeusBaseGPUError)
+            raise exception_class(e.msg) from e
+
+    def sync(self, index: int) -> None:
+        """Synchronizes the specified GPU, ensuring all previous commands have been completed."""
+        raise ZeusNotSupportedGPUError(
+            "Synchronization is not supported for AMD GPUs yet"
+        )
+
+    def __del__(self) -> None:
+        """Shuts down the AMD GPU monitoring library to release resources and clean up."""
+        with contextlib.suppress(amdsmi.AmdSmiException):
+            amdsmi.amdsmi_shut_down()  # Ignore error on shutdown. Neccessary for proper cleanup and test functionality
+
+
+_gpus = None
+
+
+def get_gpus(gpus_to_track: list[int] = None, ensure_homogeneous: bool = True) -> GPUs:
```

Comment:
What if someone calls
```python
get_gpus([1])
```
and then calls the following somewhere else?
```python
get_gpus([0, 2])
```

I think it'll return an instance that just holds GPU index 1. So this an error-prone API.

I wonder what the use case of being able to specify `gpu_indices` is. Why not just remove that option and always initialize all visible GPUs (constrained by `CUDA_VISIBLE_DEVICES`)?
The user anyway has to pass in `index` when calling GPU management APIs on `GPUs`. If they know the specific GPU indices they would like to use, they'll set `index` when calling methods. And the `index` of GPUs the user does not care about will just not be passed in. Then some GPU handles will never be used after being initialized, but it's negligible. WDYT?

Reply:
Yes, that's how it was initally done. But when the unit tests ran, it looks like they only created mock handles for gpus that were listed in gpu_indices:

```
def mock_pynvml(nvml_indices: list[int], archs: list[int]) -> None:
        assert len(nvml_indices) == len(archs)
        index_to_handle = {i: f"handle{i}" for i in nvml_indices}
        handle_to_arch = {f"handle{i}": arch for i, arch in zip(nvml_indices, archs)}
        pynvml_mock.nvmlDeviceGetCount.return_value = NUM_GPUS
        pynvml_mock.nvmlDeviceGetHandleByIndex.side_effect = lambda index: index_to_handle[index]
        pynvml_mock.nvmlDeviceGetArchitecture.side_effect = lambda handle: handle_to_arch[handle]
```
        
So `index_to_handle` would repeatedly encounter a key error. I thought it was incorrect behavior to initialize a GPU that is never used, so I implemented it with gpu_indices. But this can be changed, but then the testing code needs to be modified to provide handles for all gpus.

Reply:
Existing testing code should not stop us from implementing the correct API. Please fix the tests.